clients/zshutdown_notify/zshutdown_notify.8
.\"	$Id: d92fafc9b9c6ad361c78100478dcd7b026411e15 $
.\"
.\" Copyright 1988 by the Massachusetts Institute of Technology
.\" All rights reserved.  The file /usr/include/zephyr/mit-copyright.h
.\" specifies the terms and conditions for redistribution.
.\"
.TH ZSHUTDOWN_NOTIFY 8 "July 8, 1988" "MIT Project Athena"
.ds ]W MIT Project Athena
.SH NAME
zshutdown_notify \- notify users of impending shutdown via Zephyr
.SH SYNOPSIS
.B zshutdown_notify
.SH DESCRIPTION
.I zshutdown_notify
reads the standard input until EOF is reached, and then sends a Zephyr
notice with the official hostname, the collected input, and a warning
message as the message body of the notice.
The notice is sent to class FILSRV, instance 
.I <hostname>,
recipient "*", where <hostname> is the official hostname of the sending host.
.PP
Any users who are using resources on the sending host are expected to
subscribe to these messages so they will be warned of any impending
shutdowns.
.PP
.I zshutdown_notify
is usually invoked from within
.I /etc/shutdown.
.SH SEE ALSO
attach(1), zephyr(1), zwgc(1), shutdown(8), zhm(8), zephyrd(8)
.br
Project Athena Technical Plan Section E.4.1, `Zephyr Notification
Service'
.SH AUTHOR
.PP
C. Anthony Della Fera (Digital Equipment Corporation-Project Athena)
.sp
.SH RESTRICTIONS
Copyright (c) 1987,1988 by the Massachusetts Institute of Technology.
All Rights Reserved.
.br
.I zephyr(1)
specifies the terms and conditions for redistribution.


debian/changelog
zephyr (3.1.2-deepin1) unstable; urgency=medium

  * Rebuild

 -- Deepin Packages Builder <packages@deepin.org>  Mon, 14 Nov 2022 17:49:38 +0800

zephyr (3.1.2-1) unstable; urgency=low

  * New upstream version.

 -- Karl Ramm <kcr@debian.org>  Mon, 28 Oct 2013 10:57:21 -0400

zephyr (3.1~beta0-1) unstable; urgency=low

  * New upstream interim release.

 -- Karl Ramm <kcr@debian.org>  Mon, 08 Jul 2013 01:50:20 -0400

zephyr (3.0.2-2) unstable; urgency=low

  * Oops, you need to have automake installed if you want to use
    autoreconf.  Do so.  Closes: #658192

 -- Karl Ramm <kcr@debian.org>  Tue, 31 Jan 2012 23:09:50 -0500

zephyr (3.0.2-1) unstable; urgency=low

  * New upstream release.
  * Actually insulate the postinst from the rc script failing.
    Closes: #638260
  * Install a Danish debconf tranlation, thanks to Joe Dalton.
    Closes: #621365
  * Update Standards-Version.

 -- Karl Ramm <kcr@debian.org>  Mon, 30 Jan 2012 11:18:54 -0500

zephyr (3.0.1-1) unstable; urgency=low

  * New upstream release.
  * Update Standards-Version.
  * Define MAXPATHLEN in lib/ZLocations.c if it isn't. Closes: #508522

 -- Karl Ramm <kcr@debian.org>  Sun, 06 Mar 2011 12:54:18 -0500

zephyr (3.0-1) unstable; urgency=low

  * Final upstream 3.0 release.
  * Depend on libreadline-dev.  Closes: #553890
  * Fix problem removing obsolete conffile.  Closes: #548360
  * 3.0 (quilt) source format.
  * Add lintian overrides for the soname/packagname mismatch warnings.
  * Bump Standards-Version.

 -- Karl Ramm <kcr@debian.org>  Sun, 29 Nov 2009 12:41:37 -0500

zephyr (3.0~rc.2544-1) unstable; urgency=low

  * New upstream.
  * Supplementary readme cleanup.
  * Drop *-krb packages for real. Closes: #535099
  * GC obsolete conffiles; Closes: #538028
  * Up Standards-Version.
  
 -- Karl Ramm <kcr@debian.org>  Wed, 02 Sep 2009 14:41:48 -0400

zephyr (3.0~beta.2483-2) unstable; urgency=low

  * check for the extant ares_getnameinfo rather than the nonexistant
    ares_getaddrinfo so that we build with c-ares
  * Add the Finnish and Basque translations from #454608 and #456823
    respectively (thanks to Esko Arajärvi, Piarres Beobide)
    which somehow did not make it into 2.1.20070719.SNAPSHOT-1.1.
    Closes: #533564
  * Zephyr 3.0 does away with xzwrite, closes: #338519

 -- Karl Ramm <kcr@debian.org>  Sat, 20 Jun 2009 23:36:08 -0400

zephyr (3.0~beta.2483-1) unstable; urgency=low

  * Does not build with krb4 by default, Closes: #527339
  * fix some whitespace issues in the config scripts
  * the kerberos shared libraries should satisfy libzephyr-dev,
    Closes: #522694
  * Italian debconf translation from Luca Monducci, Closes: #531426
  * new upstream version

 -- Karl Ramm <kcr@debian.org>  Tue, 16 Jun 2009 20:34:26 -0400

zephyr (3.0~beta.2362-1) experimental; urgency=low

  * new upstream version

 -- Karl Ramm <kcr@debian.org>  Sat, 28 Mar 2009 17:52:06 -0400

zephyr (3.0~alpha.20090316-1) unstable; urgency=low

  * Kerberos 5 client support; new upstream
  * GNU config automated update: config.sub     (20070118 to 20080116),
    config.guess     (20070306 to 20080123)
  * Upstream merged CMU zwgcplus;  Closes: #207362
  * Upstream now uses c-ares: Closes: #477047
  * Fix braino in init script.  Closes: #448459
  
 -- Karl Ramm <kcr@debian.org>  Mon, 16 Mar 2009 19:57:49 -0400

zephyr (2.1.20070719.SNAPSHOT-2) unstable; urgency=low

  * Z_SendLocation could be fooled by the USER_LOGIN message (which has
    the same UID as the SERVACK we're really waiting for) arriving first.
    Wait for the SERVACK explicitly, so the USER_LOGIN gets left on the
    queue even if it arrives early. (thanks, Mark!)
  * Acknowledge NMUs
  * Build-depend on x11proto-core-dev.  Closes: #515456
  * Japanese debconf translation, thanks to Hideki Yamane.  Closes: #510734

 -- Karl Ramm <kcr@debian.org>  Sun, 01 Mar 2009 02:52:20 -0500

zephyr (2.1.20070719.SNAPSHOT-1.2) unstable; urgency=low

  * Non-maintainer upload.
  * Fix pending l10n issues. Debconf translations:
    - Swedish. Closes: #491531
  * Brazilian Portuguese. Closes: #501262

 -- Christian Perrier <bubulle@debian.org>  Mon, 06 Oct 2008 07:16:54 +0200

zephyr (2.1.20070719.SNAPSHOT-1.1) unstable; urgency=low

  * Non-maintainer upload to fix pending l10n issues
  * Debconf templates and debian/control reviewed by the debian-l10n-
    english team as part of the Smith review project. Closes: #453291
  * [Debconf translation updates]
    - Galician. Closes: #454511, #447936
    - Finnish. Closes: #454608
    - Czech. Closes: #455240
    - Spanish. Closes: #455576
    - Italian. Closes: #456225
    - Basque. Closes: #456823
    - German. Closes: #456981
    - Dutch. Closes: #457034, #447724
    - Russian. Closes: #457087, #451807
    - Portuguese. Closes: #457111
    - Vietnamese. Closes: #457301
    - French. Closes: #457560, #435494, #435495

 -- Christian Perrier <bubulle@debian.org>  Wed, 20 Feb 2008 20:11:27 +0100

zephyr (2.1.20070719.SNAPSHOT-1) unstable; urgency=low

  * new upstream, Closes: #433603
  * byteswap port number for display in zwgc, Closes: #243101
  * stick newline after errant error message, Closes: #242839
  * wait for zhm to die when restarting. Closes: #41419

 -- Karl Ramm <kcr@debian.org>  Fri, 20 Jul 2007 13:23:48 -0400

zephyr (2.1.20010518.SNAPSHOT-18) unstable; urgency=low

  * fix braino in postinst.  Closes: #433953
  * Don't mark internal use messages as translatable.  Closes: #413237
  * Galician translation.  Closes: #412802
  * Dutch translation.  Closes: #413876  
  * Portugese translation.  Closes: #414817
  * Updates Spanish translatoin.  Closes: #420839
  
  * Package supports using hesiod to find zephyr servers.  Closes: #174535
  * Package does not fail to configure if the daemon won't start.  Closes: #120421
  * zhm shouldn't hang for ever if it can't find a server on startup.  Closes: #433602

  * remove DH_COMPAT setting and update substvars to placate lintian.


clients/zshutdown_notify/zshutdown_notify.c
/* This file is part of the Project Athena Zephyr Notification System.
 * It contains code for "zshutdown_notify", a utility called by
 * shutdown(8) to do Zephyr notification on shutdown.
 *
 *	Created by:	C. Anthony Della Fera
 *
 *	$Id: 2f51cc05b9f2db41c73282c5f50cb25bf9c2ce89 $
 *
 *	Copyright (c) 1987, 1993 by the Massachusetts Institute of Technology.
 *	For copying and distribution information, see the file
 *	"mit-copyright.h".
 */

#include <sysdep.h>
#include <zephyr/mit-copyright.h>
#include <zephyr/zephyr.h>

#include <sys/socket.h>
#include <netdb.h>
#include <arpa/nameser.h>

#ifndef lint
static const char rcsid_zshutdown_notify_c[] =
    "$Id: 2f51cc05b9f2db41c73282c5f50cb25bf9c2ce89 $";
#endif

#define N_KIND		UNSAFE
#define N_CLASS		"FILSRV"
#define N_OPCODE	"SHUTDOWN"
#define N_DEF_FORMAT	"From $sender:\n@bold(Shutdown message from $1 at $time)\n@center(System going down, message is:)\n\n$2\n\n@center(@bold($3))"
#define N_FIELD_CNT	3

#ifdef HAVE_KRB4
#define SVC_NAME	"rcmd"
#endif

/*
 * Standard warning strings appended as extra fields to
 * the message body.
 */

static char warning[] = "Please detach any filesystems you may have\nattached from this host by typing detach -host %s";

/*ARGSUSED*/
int
main(int argc,
     char *argv[])
{
    ZNotice_t notice;
    struct hostent *hp;
    int retval;
    char hostname[NS_MAXDNAME];
    char msgbuff[BUFSIZ], message[Z_MAXPKTLEN], *ptr;
    char scratch[BUFSIZ];
    char *msg[N_FIELD_CNT];
#ifdef HAVE_KRB4
    char tkt_filename[MAXPATHLEN];
    char rlm[REALM_SZ];
    char hn2[NS_MAXDNAME];
    char *cp;
#endif

    if (gethostname(hostname, sizeof(hostname)) < 0) {
	com_err(argv[0], errno, "while finding hostname");
	exit(1);
    }

    if ((hp = gethostbyname(hostname)) != NULL)
	    (void) strcpy(hostname, hp->h_name);

    msg[0] = hostname;
    msg[1] = message;
    sprintf(scratch, warning, hostname);
    msg[2] = scratch;

#ifdef HAVE_KRB4
    (void) sprintf(tkt_filename, "/tmp/tkt_zshut_%d", getpid());
    krb_set_tkt_string(tkt_filename);

    cp = krb_get_phost(hostname);
    if (cp)
	(void) strcpy(hn2, cp);
    else {
	fprintf(stderr, "%s: can't figure out canonical hostname\n",argv[0]);
	exit(1);
    }
    retval = krb_get_lrealm(rlm, 1);
    if (retval) {
	fprintf(stderr, "%s: can't get local realm: %s\n",
		argv[0], krb_get_err_text(retval));
	exit(1);
    }
    retval = krb_get_svc_in_tkt(SVC_NAME, hn2, rlm,
				SERVER_SERVICE, SERVER_INSTANCE, 1,
				(char *)KEYFILE);
    if (retval) {
	fprintf(stderr, "%s: can't get tickets: %s\n",
		argv[0], krb_get_err_text(retval));
	exit(1);
    }
#endif

    if ((retval = ZInitialize()) != ZERR_NONE) {
	com_err(argv[0], retval, "while initializing");
	exit(1);
    }

    ptr = message;

    for (;;) {
	if (!fgets(msgbuff, sizeof(msgbuff), stdin))
	    break;
	if ((strlen(msgbuff) + (ptr - message)) > Z_MAXPKTLEN){
	    break;
	}
	(void) strcpy(ptr, msgbuff);
	ptr += strlen(ptr);
    }

    (void) memset((char *)&notice, 0, sizeof(notice));

    notice.z_kind = N_KIND;
    notice.z_port = 0;
    notice.z_charset = ZCHARSET_UNKNOWN;
    notice.z_class = N_CLASS;
    notice.z_class_inst = hostname;
    notice.z_opcode = N_OPCODE;
    notice.z_sender = 0;
    notice.z_message_len = 0;
    notice.z_recipient = "";
    notice.z_default_format = N_DEF_FORMAT;

    retval = ZSendList(&notice, msg, N_FIELD_CNT, ZAUTH);
#ifdef HAVE_KRB4
    (void) dest_tkt();
#endif

    if (retval != ZERR_NONE) {
	com_err(argv[0], retval, "while sending notice");
	exit(1);
    }
    return 0;
}
